jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathResponseListener.java

```diff
@@ -0,0 +1,192 @@
+    private Path path;
+    private Response response;
+    private Throwable failure;
+    private FileOutputStream fileOut;
+    private FileLock fileLock;
+    
+    public PathResponseListener(Path path) throws FileNotFoundException, IOException
+    {   
+        if (!path.isAbsolute())
+        {
+            throw new FileNotFoundException();
+        }
+        
+        this.path = path;
+        
+        try 
+        {
+            fileOut = new FileOutputStream(this.path.toFile(), true);
+            fileOut.getChannel().truncate(0);
+            fileLock = fileOut.getChannel().lock();
+        }
+        catch (IOException e) 
+        {   
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to instantiate object", e);
+            else
+                throw e;
+        }
+    }
+
+    @Override
+    public void onContent(Response response, ByteBuffer content) throws IOException
+    {
+        if (response.getStatus() != HttpStatus.OK_200)
+        {
+            throw new HttpResponseException(String.format("HTTP status code of this response %d", response.getStatus()), response);
+        }
+        
+        if (!fileLock.isValid())
+        {
+            throw new IOException("File lock is not valid");
+        }
+        try
+        {   
+            fileOut.getChannel().write(content);
+        }
+        catch (IOException e) 
+        {   
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to write file", e);
+            else
+                throw e;
+        }
+    }
+
+    @Override
+    public void onComplete(Result result)
+    {
+        if (result.isFailed() && this.failure == null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Result failure", failure);
+        }
+        
+        this.response = result.getResponse();
+        this.completable.complete(this.path);
+        
+        try
+        {
+            fileLock.close();
+        }
+        catch (IOException e)
+        {   
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to close file", e);
+        }
+    }
+    
+    public Response get(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException, ExecutionException 
+    {
+        boolean expired = !responseLatch.await(timeout, unit);
+        if (expired && this.response == null)
+            throw new TimeoutException();
+        try (AutoLock ignored = lock.lock())
+        {
+            // If the request failed there is no response.
+            if (response == null)
+                throw new ExecutionException(failure);
+            return response;
+        }
+    }
+    
+    public Response get() throws InterruptedException, ExecutionException
+    {
+        this.completable.get();
+        return this.response;
+    }
+   
+    public static CompletableFuture<Path> write(Request request, Path path)
+    {
+        return CompletableFuture.supplyAsync(() -> 
```

Comment:
This method should be implemented in terms of `PathResponseListener`, not by rewriting the same logic again.

Reply:
I'm a little bit confused with this requirements as I'm not clear what should be done exactly.

In [comment](https://github.com/jetty/jetty.project/issues/929#issuecomment-2326479685) you asked to implement static method `write` which returns `CompletableFuture<Path>` object.  

As we have a `Request` in parameters, we can't get `ByteBuffer` object as it is done in `onContent` so we should use `InputStreamResponseListener` to get bytes stream - that's why separate logic was introduced in the static `write` method.

Reply:
@arsenalzp this could be implemented in this way:

```java
public static CompletableFuture<Path> write(Request request, Path path) {
  PathResponseListener l = new PathResponseListener(path);
  request.send(l);
  return l;
}
```

Reply:
I was thinking about it, however this approach is bringing to many casts.
However, I did as you recommend!

Reply:
@arsenalzp in your current work, there is a clash between returning `CompletableFuture<Path>` or `CompletableFuture<Response>`.

Perhaps the best way is to return both 😃 

You can create a simple `record PathResponse(Response response, Path path) {}` and return `CompletableFuture<PathResponse>`.

Reply:
I'm fine with cast, don't want to make code more complicated. So, if we decided to return CompletableFuture<Path> then let's return it.
Is current version OK?